src/bgw/scheduler.c

```diff
@@ -0,0 +1,402 @@
+		return 0;
+
+	TimestampDifference(GetCurrentTimestamp(), by_time, &timeout_sec, &timeout_usec);
+
+
+	if (timeout_sec <= 0 && timeout_usec <= 0)
+		return 0;
+
+	return timeout_sec * 1000L + timeout_usec / 1000;
+}
+
+static void
+scheduled_bgw_job_start_worker(ScheduledBgwJob *sjob)
+{
+
+	pid_t		pid;
+	BgwHandleStatus status;
+
+
+	Assert(sjob->enabled);
+
+	StartTransactionCommand();
+	ts_bgw_job_stat_mark_start(sjob->job->fd.id);
+	if (ts_bgw_job_has_timeout(sjob->job))
+		sjob->timeout_at = ts_bgw_job_timeout_at(sjob->job, GetCurrentTransactionStartTimestamp());
+	else
+		sjob->timeout_at = DT_NOEND;
+	CommitTransactionCommand();
+
+	elog(DEBUG1, "scheduling job %d \"%s\"", sjob->job->fd.id, NameStr(sjob->job->fd.job_name));
+
+
+	sjob->handle = ts_bgw_job_start_worker(sjob->job);
+	status = WaitForBackgroundWorkerStartup(sjob->handle, &pid);
+	switch (status)
+	{
+		case BGWH_POSTMASTER_DIED:
+			bgw_scheduler_on_postmaster_death();
+			break;
+		case BGWH_STARTED:
+			/* all good */
+			break;
+		case BGWH_STOPPED:
+			sjob->handle = NULL;
+			StartTransactionCommand();
+			ts_bgw_job_stat_mark_end(sjob->job, false);
+			scheduled_bgw_job_mark_stopped(sjob);
+			CommitTransactionCommand();
+			break;
+		case BGWH_NOT_YET_STARTED:
+			/* should not be possible */
+			elog(ERROR, "unexpected bgworker state %d", status);
+			break;
+	}
+}
+
+static ScheduledBgwJob *
+scheduled_bgw_job_create(BgwJob *job, MemoryContext mctx)
+{
+	BgwJobStat *job_stat = ts_bgw_job_stat_find(job->fd.id);
+	ScheduledBgwJob *sjob = MemoryContextAlloc(mctx, sizeof(ScheduledBgwJob));
+
+	*sjob = (ScheduledBgwJob)
+	{
+		.job = job,
+			.enabled = ts_bgw_job_stat_should_execute(job_stat, job),
+			.next_start = ts_bgw_job_stat_next_start(job_stat, job),
+	};
+
+	return sjob;
+}
+
+static bool
+scheduled_bgw_job_is_running(ScheduledBgwJob *sjob)
+{
+	return sjob->handle != NULL;
+}
+
+static void
+scheduled_bgw_job_mark_stopped(ScheduledBgwJob *sjob)
+{
+	BgwJobStat *job_stat = ts_bgw_job_stat_find(sjob->job->fd.id);
+
+	if (!ts_bgw_job_stat_end_was_marked(job_stat))
+	{
+		/*
+		 * This happens not on a crash but rather upon a signal (cancel or
+		 * terminate) of the job
+		 */
+		ts_bgw_job_stat_mark_end(sjob->job, false);
+		/* refresh */
+		job_stat = ts_bgw_job_stat_find(sjob->job->fd.id);
+	}
+
+	sjob->handle = NULL;
+	sjob->next_start = ts_bgw_job_stat_next_start(job_stat, sjob->job);
+	sjob->enabled = ts_bgw_job_stat_should_execute(job_stat, sjob->job);
+}
+
+static
+List *
+scheduled_bgw_job_list_create(MemoryContext mctx)
+{
+	List	   *res = NIL;
+	List	   *jobs = ts_bgw_job_get_all(mctx);
+	ListCell   *lc;
+	MemoryContext old;
+
+	foreach(lc, jobs)
+	{
+		BgwJob	   *job = lfirst(lc);
+		ScheduledBgwJob *sjob = scheduled_bgw_job_create(job, mctx);
+
+		old = MemoryContextSwitchTo(mctx);
+		res = lappend(res, sjob);
+		MemoryContextSwitchTo(old);
+	}
+	return res;
+}
+
+static
+bool
+scheduled_bgw_job_list_start_workers()
```

Comment:
I keep reading the function as scheduled_bgw_job_list_start**ed**_workers, but I don't know a better thing to name it. Adding a double underscore between the noun and verb, as in `scheduled_bgw_job_list__start_workers` would help, if we were willing to adopt that convention.

Reply:
I don't think we want to adopt that convention let's think about better names. Would `scheduled_bgw_job_list_launch_workers` be better?

Reply:
Yes, I think it's easier to understand.